### ISSUE TITLE
twisted.python.lockfile.FilesystemLock.lock fails with EEXIST

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3367]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3367|
|Type|defect|
|Created|2008-07-24 16:22:44Z|
|Branch|https://github.com/twisted/twisted/tree/lockfile-3367|

If the lock is acquired after `lock` handles the `ENOENT` from `readlink` but before it gets to the `symlink` call, the `symlink` call will fail with `EEXIST`.

<details><summary>Searchable metadata</summary>

```
trac-id__3367 3367
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__branches_lockfile_3367 branches/lockfile-3367
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1216916564000000 1216916564000000
changetime__1217851450000000 1217851450000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [24341]) Branching to 'lockfile-3367'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

Ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_3"></a>|@washort set owner to @exarkun|
|-|-|

Nothing sets FilesystemLock.locked to True anymore. Nothing checks it either. Maybe it should be removed?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Yea, it seems like it's of marginal value.  Maybe we should deprecate it, but I guess I expand the tests and fix the behavior to match the old behavior for now, since it was a public attribute.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [24385]) test and fix the locked attribute

refs [#3367](https://github.com/twisted/twisted/issues/3367)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_7"></a>|@washort set owner to @exarkun|
|-|-|

Looks good, merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [24387]) Handle ENOENT from rmlink

refs [#3367](https://github.com/twisted/twisted/issues/3367)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [24388]) mock kill to make the results deterministic

refs [#3367](https://github.com/twisted/twisted/issues/3367)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun removed owner|
|-|-|

I noticed some more problems after the review. :(

The code is now slightly more robust on POSIX and a lot more robust and functional on Windows.

Build results:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/lockfile-3367

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to @exarkun|
|-|-|

OK, crazy but great. Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [24508]) Merge lockfile-3367

Author: exarkun
Reviewer: therve
Fixes: [#3367](https://github.com/twisted/twisted/issues/3367)

Fix a concurrency issue with `FilesystemLock` on POSIX where the `lock`
method could end up raising `OSError(EEXIST)`.  Windows support is also
greatly enhanced: most behavior should be the same as on POSIX, including
detection of stale locks held by processes which no longer exist.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
